### DISCUSSION TITLE
Cannot jump to the declaration of a symbol

### DISCUSSION THREAD

**Discussion by halloleo** [joaotavora/eglot#739] [Q&A] [answered]
In a Python file like this

```
from muddle_funcs import muddle_me, demuddle_me
.
.
.
demuddle_me(input, output, muddler, dont_fail)
.
.
.
```

I have the cursor on the function call to `demuddle_me`, then I execute `eglot-find declaration` and I expect to jump to the file `muddle_funcs.py` where `demuddle_me` is defined.  muddle_funcs.py  is in the same directory as the present file.

Emacs returns:

    [eglot] Sorry, this server doesn't do :textDocument/declaration

I have pyls installed so what goes wrong?

**Comment by joaotavora** [comment]
Thanks for starting this conversation.  Can you make a small zip or archive or the toy project you're tying this and post it here.  It would help.  My python chops come and go, it would help to have a sample around and not have to type anything.

**Comment by nemethf** [comment, marked as answer]
I'm afraid Eglot is right in this case.  Pyls does not implement textDocument/declaration. See its [hookspecs](https://github.com/palantir/python-language-server/blob/develop/pyls/hookspecs.py).  Additionally pyls is more or less abandoned.  Pylsp is the new pyls, but there seems to be no difference in this regard.  (Well, pylsp is almost the same as pyls.)

Have you tried xref-find-definitions or xref-find-references? Maybe those help you jump to the point of the declaration.

**Reply by halloleo** [reply]
`xref-find-definitions` works! Thanks.

Interestingly it use _something_ from eglot, because `xref-find-definitions` does not work without `eglot-ensure` in the `python-mode` hook.

What's that? And is `xref-find-definitions` the preferred way in eglot to find a symbol?
